src/plugins/interaction/GraphInteraction/GraphInteraction.cpp

```diff
@@ -77,112 +88,122 @@ std::ifstream& GotoLine(std::ifstream& file, unsigned int num) {
-                edge_counter++;
-                int id_start = std::stoi(words[0]), id_end = std::stoi(words[1]);
-                double length = std::stod(words[2]);
-
-                bool edge_nodes_exist = nodes.count(id_start) > 0 && nodes.count(id_end) > 0; // reality check
-                if (edge_nodes_exist) {
-                    auto edge_ptr = graph_msg->data.add_edges();
-                    edge_ptr->set_start_node_id(id_start);
-                    edge_ptr->set_end_node_id(id_end);
-                    edge_ptr->set_weight(length);
-
-                    if (found_labels) {
-                        // Get Street Name
-                        GotoLine(names_file, edge_counter);
-                        std::string street_name;
-                        std::getline(names_file, street_name);
-                        edge_ptr->set_label(street_name);
-                    }
-
-                    // Visualize the Edges
-                    if (vis_graph_) {
-                        auto edge_shape = std::make_shared<scrimmage_proto::Shape>();
-                        edge_shape->set_persistent(true);
-                        edge_shape->set_opacity(1.0);
-                        scrimmage::set(edge_shape->mutable_color(), 255, 0, 0);
-                        scrimmage::set(edge_shape->mutable_line()->mutable_start(), nodes[id_start]);
-                        scrimmage::set(edge_shape->mutable_line()->mutable_end(), nodes[id_end]);
-                        draw_shape(edge_shape);
-                    }
-                }
-            }
-        }
-        pub_graph_->publish(graph_msg);
+    std::pair<boost::adjacency_list<>::vertex_iterator,
+        boost::adjacency_list<>::vertex_iterator> vs = boost::vertices(g_);
+
+    std::map<uint64_t, Eigen::Vector3d> nodes;
+    std::map<uint64_t, uint64_t> boost_vert_to_osmid;
+
+    auto graph_msg = std::make_shared<sc::Message<sm::Graph>>();
+    graph_msg->data.set_id(id_);
+    for (; vs.first != vs.second; ++vs.first) {
+        // double longitude = boost::get(&NodeProperties::x, g_, *vs.first);
+        // double latitude = boost::get(&NodeProperties::y, g_, *vs.first);
+        // uint64_t this_osmid = boost::get(&NodeProperties::osmid, g_, *vs.first);
+        // NOTE: the above three lines are equivalent to the below three lines
+        double longitude = g_[*vs.first].x;
+        double latitude = g_[*vs.first].y;
+        double this_osmid = g_[*vs.first].osmid;
+        boost_vert_to_osmid[*vs.first] = this_osmid;
+
+        double x, y, z;
+        parent_->projection()->Forward(latitude, longitude, 0.0, x, y, z);
+        nodes[this_osmid] = Eigen::Vector3d(x, y, z);
+        auto node_ptr = graph_msg->data.add_nodes();
+        node_ptr->set_id(this_osmid);
+        sc::set(node_ptr->mutable_point(), nodes[this_osmid]);
+    }
 
-        // Visualize the Nodes
+    auto es = boost::edges(g_);
+    for (; es.first != es.second; ++es.first) {
```

Comment:
can we initialize this variable in the loop?

Reply:
Yes, good idea. Thanks!